### PR TITLE
Project keystone

### PR DESCRIPTION
**Pull request by cindylovescoding**
**Design Logic:**
A configuration “keystoneSolutionConfig” has been added in support-topic.service.ts. For each pesId, I have defined a dictionary to map the keywords to keystone detector Id:

  private keystoneSolutionConfig = {
    "14748" : [
        {
            keywords: ["tls 1.3", "tls1.3", "premium v3", "premiumv3", "p3v3", "pv3", "sendgrid", "sitecore"],
            keystoneDetectorId: "test_keystone_detector"
        }
    ]
  };

In case submission flow, we check from case subject and see we can find any mapping keystone detector from this configuration. If so, we render keystone solution view, see more in "**Testing scenarios in case submission flow:**" below
If not, we render the original view for mapped analysis/detector/dynamic analysis

This is designed with this config so that we can power keystone solution from a keystone detector with more flexibility, but also not blocking to render the original diagnostic view by waiting for the response from keystone detector first and then decide which view to load. 

**Testing scenarios in case submission flow:**
You can test the case submission flow in staging environment:  
https://ms.portal.azure.com/?websitesextension_ext=asd.env%3Dstaging#home

Try to submit a case with any of these substrings (["tls 1.3", "tls1.3", "premium v3", "premiumv3", "p3v3", "pv3", "sendgrid", "sitecore"]) in your case subject.

**1. Keystone rendering**
image

**2. Keystone solution within Analysis view**
image

**3. Keystone solution within Detector view**
image

**4. Keystone solution within Dynamic Analysis view**
image

 Logging/Accessibility has been taken care of as well in this PR.

### CODE REVIEW

File: AngularApp/projects/app-service-diagnostics/src/app/shared/shared.module.ts [hunk 1]
```diff
@@ -202,7 +202,7 @@ import { RiskAlertsPanelComponent } from './components/risk-alerts-panel/risk-al
         CrashMonitoringAnalysisComponent,
         CollapsibleListFabricComponent,
         RiskAlertsNotificationComponent,
-        RiskAlertsPanelComponent
+        RiskAlertsPanelComponent,
```
Comment (ShekharGupta1988) [inline AngularApp/projects/app-service-diagnostics/src/app/shared/shared.module.ts:205]: get rid of ,

File: AngularApp/projects/app-service-diagnostics/src/app/shared-v2/services/support-topic.service.ts [hunk 2]
```diff
@@ -23,6 +23,15 @@ export class SupportTopicService {
     "15551": ["32632390","32632389","32632430","32632396","32632397","32632401","32632431","32632436","32632434","32632406","32632389","32632390","32632430","32632393","32632398","32632403","32632409","32632413","32632414","32632415","32632385","32632418","32632419","32632422","32632424","32632438","32632399","32632408","32632421","32632426","32632405","32683732","32632427","32632386","32632387","32632388","32632395","32632404","32632416","32632402","32632420","32632425","32632428","32632432","32632437","32632407","32740235","32740236","32740237","32740238","32740239","32740240","32740234"]
   };
 
+  private keystoneSolutionConfig = {
+    "14748" : [
+        {
+            keywords: ["tls 1.3", "tls1.3", "premium v3", "premiumv3", "p3v3", "pv3", "sendgrid", "sitecore"],
```
Comment (ShekharGupta1988) [inline AngularApp/projects/app-service-diagnostics/src/app/shared-v2/services/support-topic.service.ts:29]: Just FYI for other reviewers - 
This config is no longer needed and Cindy will be updating the PR to remove this.
The only thing needed here is the detector Id.

File: AngularApp/projects/app-service-diagnostics/src/app/shared/shared.module.ts (hunk 1 above)
Comment (cindylovescoding) [reply AngularApp/projects/app-service-diagnostics/src/app/shared/shared.module.ts]: Done!

File: AngularApp/projects/app-service-diagnostics/src/app/shared-v2/services/support-topic.service.ts (hunk 2 above)
Comment (cindylovescoding) [reply AngularApp/projects/app-service-diagnostics/src/app/shared-v2/services/support-topic.service.ts]: > Just FYI for other reviewers -
> This config is no longer needed and Cindy will be updating the PR to remove this.
> The only thing needed here is the detector Id.

PR has been updated. Logic has been changed to:
1. Frontend reads resource configuration for each resource type to get keystone detector Id
2. Frontend make the request to keystone detector id (from step 1) with query parameter "extractKeywords=true"
3. Frontend gets keywords set from step 2, and compare the current case subject with the keywords set, see if the case subject contains any of the keywords
4. For step 3, if true, Frontend makes two get detector requests (one for keystone detector, one for support topic id matched detector) to render keystone solution view
5. For step 3, if false, Frontend makes only one get detector request to get the support topic id matched detector and render it in the original view

**Comment by cindylovescoding** [comment]
Discussed with Shekhar and Khaled, improved the view a little more:
1. Remove closing blade notification banner at the top
2. Move the timer picker into section **option 2** inside the original detector/analysis view

image

File: AngularApp/projects/app-service-diagnostics/src/app/resources/web-sites/services/web-sites.service.ts [hunk 3]
```diff
@@ -81,6 +81,17 @@ export class WebSitesService extends ResourceService {
         }));
     }
 
+    public getKeystoneDetectorId(): Observable<string>{
+        return this.warmUpCallFinished.pipe(flatMap(() => {
+            if (this.appType == AppType.WebApp && this.platform == OperatingSystem.windows) {
```
Comment (ShekharGupta1988) [inline AngularApp/projects/app-service-diagnostics/src/app/resources/web-sites/services/web-sites.service.ts:86]: I think we should enable this detector for all app types on windows, except functions app.

File: AngularApp/projects/app-service-diagnostics/src/app/shared-v2/services/support-topic.service.ts [hunk 4]
```diff
@@ -1,73 +1,108 @@
...
-                  detectorPath = `/analysis/searchResultsAnalysis/search`;
-              }
-            }    
-            return {path: detectorPath, queryParams: {"searchTerm": searchTerm}};
-          }));
-    }));
-  }
+
+                let keywordsList = [];
+                return this._webSiteService.getKeystoneDetectorId().pipe(flatMap(keystoneDetectorId => {
```
Comment (ShekharGupta1988) [inline AngularApp/projects/app-service-diagnostics/src/app/shared-v2/services/support-topic.service.ts:74]: this _websiteService got me little confused :)
This variable is actually of type Resource Service (so the right resource will get injected), just that the name of the variable is _websiteService. 

Can you update the variable name to be _resourceService

File: AngularApp/projects/app-service-diagnostics/src/app/resources/web-sites/services/web-sites.service.ts (hunk 3 above)
Comment (cindylovescoding) [reply AngularApp/projects/app-service-diagnostics/src/app/resources/web-sites/services/web-sites.service.ts]: I have update the PR for this fix

File: AngularApp/projects/app-service-diagnostics/src/app/shared-v2/services/support-topic.service.ts (hunk 4 above)
Comment (cindylovescoding) [reply AngularApp/projects/app-service-diagnostics/src/app/shared-v2/services/support-topic.service.ts]: Fixed it already:)

File: AngularApp/projects/app-service-diagnostics/src/app/resources/web-sites/services/web-sites.service.ts [hunk 5]
```diff
@@ -81,6 +81,17 @@ export class WebSitesService extends ResourceService {
         }));
     }
 
+    public getKeystoneDetectorId(): Observable<string>{
+        return this.warmUpCallFinished.pipe(flatMap(() => {
+            if (this.appType !== AppType.FunctionApp && this.platform == OperatingSystem.windows) {
+                return of("test_keystone_detector");
```
[the remark this answers is not in the document]
Comment (cindylovescoding) [reply AngularApp/projects/app-service-diagnostics/src/app/resources/web-sites/services/web-sites.service.ts:87]: That is right Xiaoxu

**Review by ShekharGupta1988 — APPROVED** [review]

**Review by XiaoxuHU — APPROVED** [review]

**Pull request merged**